include/io/mdarray.hpp

```diff
@@ -0,0 +1,274 @@
+#ifndef NGEN_IO_MDARRAY_HPP
+#define NGEN_IO_MDARRAY_HPP
+
+#include <initializer_list>
+#include <stdexcept>
+#include <vector>
+
+namespace io {
+
+template<typename T>
+class mdarray
+{
+  public:
+    using value_type      = T;
+    using container_type  = std::vector<value_type>;
+    using ilist           = std::vector<std::size_t>;
+    using size_type       = typename container_type::size_type;
+    using difference_type = typename container_type::difference_type;
+    using reference       = typename container_type::reference;
+    using const_reference = typename container_type::const_reference;
+    using pointer         = typename container_type::pointer;
+    using const_pointer   = typename container_type::const_pointer;
+
+    struct iterator;
+    friend iterator;
+
+    // Deleted default constructor. mdarray must be given a rank at least.
+    mdarray() = delete;
+
+    mdarray(size_type rank) noexcept
+        : m_shape(rank)
+        , m_data() {};
+
+    mdarray(ilist dsizes)
+        : m_shape(dsizes.begin(), dsizes.end())
+        , m_data(this->max_size()) {};
+
+    /**
+     * Retrieve a reference to the value at the given index.
+     *
+     * @example
+     * mdarray<int> x(2); // rank 2
+     * x.at({0, 0});      // get the value at (0, 0) in a 2D array.
+     * x.at({1, 1}) = 3;  // set the value at (1, 1) in a 2D array.
+     *
+     * @param n Index list
+     * @return reference 
+     */
+    reference at(const ilist& n)
+    {
+        return this->m_data.at(this->index(n));
+    }
+
+    /**
+     * Retrieve const reference to the value at the given index.
+     *
+     * @example
+     * mdarray<int> x(2); // rank 2
+     * x.at({0, 0});      // get the value at (0, 0) in a 2D array.
+     *
+     * @param n Index list
+     * @return reference 
+     */
+    const_reference at(const ilist& n) const
+    {
+        return this->m_data.at(this->index(n));
+    }
+
+    void insert(const ilist& n, value_type value)
+    {
+        for (size_type i = 0; i < this->m_shape.size(); i++) {
+            if (n[i] >= this->m_shape[i])
+                throw std::out_of_range("out of range");
+        }
+
+        size_type index = this->index(n);
+    
+        if (index > this->max_size())
+            throw std::out_of_range("out of range");
+        
+        if (index > this->size() || this->size() == 0)
+            this->m_data.resize(index + 1);
+    
+        this->m_data.at(index) = value;
```

Comment:
```suggestion
        this->m_data[index] = value;
```
There's no reason to ask the underlying `vector` to repeat the bounds-check we ensured would pass just above.

Reply:
Thanks! I always forget that `.at` does a bounds check.